### PR TITLE
Require braces to be present for environment variable expansion

### PR DESCRIPTION
**Pull request by andrewkroh**
The docs were updated in #1290.

### CODE REVIEW

File: libbeat/cfgfile/cfgfile.go [hunk 1]
```diff
@@ -103,3 +105,65 @@ func expandEnv(config []byte) []byte {
...
+// expand method was modified to only expand variables defined with braces and
+// ignore $var.
+
+// Expand replaces ${var} in the string based on the mapping function.
+func expand(s string, mapping func(string) string) string {
+	buf := make([]byte, 0, 2*len(s))
+	// ${} is all ASCII, so bytes are fine for this operation.
+	i := 0
+	for j := 0; j < len(s); j++ {
+		if s[j] == '$' && j+2 < len(s) && s[j+1] == '{' {
```
Comment (andrewkroh) [inline libbeat/cfgfile/cfgfile.go]: This is the only line that was changed from the `os.Expand` function.

File: libbeat/cfgfile/cfgfile_test.go [hunk 2]
```diff
@@ -46,21 +46,31 @@ func TestExpandEnv(t *testing.T) {
...
 
-		// Defaults can only be specified when using braces.
+		// Defaults can be specified.
 		{"x${Z:D}", "xD"},
 		{"x${Z:A B C D}", "xA B C D"}, // Spaces are allowed in the default.
 		{"x${Z:}", "x"},
 
-		// Defaults don't work unless braces are used.
-		{"x$y:D", "xy:D"},
+		// Un-matched braces are swallowed by the Go os.Expand function.
```
Comment (tsg) [inline libbeat/cfgfile/cfgfile_test.go]: This unfortunately increases the chance of it showing up accidentally, right?

File: libbeat/cfgfile/cfgfile_test.go [hunk 3]
```diff
@@ -46,21 +46,31 @@ func TestExpandEnv(t *testing.T) {
...
+		// Special environment variables are not replaced.
+		{"$*", "$*"},
+		{"${*}", ""},
+		{"$@", "$@"},
+		{"${@}", ""},
+		{"$1", "$1"},
+		{"${1}", ""},
+
+		{"", ""},
+		{"$$", "$$"},
```
Comment (tsg) [inline libbeat/cfgfile/cfgfile_test.go]: The only other thing that might be useful is some sort of escaping. e.g. `${{X}` becomes `${X}`. Would it be tricky to implement?

File: libbeat/cfgfile/cfgfile_test.go (hunk 2 above)
Comment (andrewkroh) [inline libbeat/cfgfile/cfgfile_test.go]: Yeah, it does. I'll see about modify their code to remove this behavior.

File: libbeat/cfgfile/cfgfile.go [hunk 4]
```diff
@@ -103,3 +105,65 @@ func expandEnv(config []byte) []byte {
...
+		if len(s) > 2 && isShellSpecialVar(s[1]) && s[2] == '}' {
+			return s[1:2], 3
+		}
+		// Scan to closing brace
+		for i := 1; i < len(s); i++ {
+			if s[i] == '}' {
+				return s[1:i], i + 1
+			}
+		}
+		return "", 1 // Bad syntax; just eat the brace.
```
Comment (urso) [inline libbeat/cfgfile/cfgfile.go]: we silently want to ignore bad syntax?

File: libbeat/cfgfile/cfgfile.go [hunk 5]
```diff
@@ -103,3 +105,65 @@ func expandEnv(config []byte) []byte {
...
+	case isShellSpecialVar(s[0]):
```
Comment (urso) [inline libbeat/cfgfile/cfgfile.go]: checking conditions in caller, this case is not possible anymore. We're always in first case `s[0] == '{'` => no more need for isShellSpecialVar and isAlphaNum

File: libbeat/cfgfile/cfgfile_test.go (hunk 3 above)
Comment (urso) [inline libbeat/cfgfile/cfgfile_test.go]: or optionally escape `$$` to become `$`, in this case one can write `$${X}` to get `${X}`. More common escape character would be \, but this one might be somehow interpreted by yaml parser of no quotes or double quotes are used.

**Comment by tsg** [comment]
Merging this as is, to have it in alpha1. Remaining comments can be addressed in other PRs.

**Pull request merged**